src/Agent.Listener/Configuration/AutoLogonRegistryManager.cs

```diff
@@ -0,0 +1,339 @@
+            string cmdExePath = System.Environment.GetEnvironmentVariable("comspec");
+            if (string.IsNullOrEmpty(cmdExePath))
+            {
+                cmdExePath = "cmd.exe";
+            }
+
+            //file to run in cmd.exe
+            var filePath = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Root), "run.cmd");
+            
+            //extra "" are to handle the spaces in the file path (if any)
+            var startupCommand = $@"start ""Agent with AutoLogon"" {cmdExePath} /D /S /C """"{filePath}""""";
+            
+            Trace.Info($"Setting startup command as '{startupCommand}'");
+            return startupCommand;
+        }  
+
+        private void RevertOriginalValue(RegistryHive targetHive, string subKeyName, string name)
+        {
+            var nameofTheBackupValue = GetBackupValueName(name);
+            var originalValue = _registryManager.GetValue(targetHive, subKeyName, nameofTheBackupValue);
+
+            if (string.IsNullOrEmpty(originalValue))
+            {
+                //there was no backup value present, just delete the current one
+                _registryManager.DeleteValue(targetHive, subKeyName, name);
+            }
+            else
+            {
+                //revert to the original value
+                _registryManager.SetValue(targetHive, subKeyName, name, originalValue);
+            }
+
+            //delete the value that we created for backup purpose
+            _registryManager.DeleteValue(targetHive, subKeyName, nameofTheBackupValue);
+        }
+
+        private void RevertRegistrySettingsForCurrentUser()
+        {
+            var hive = RegistryHive.CurrentUser;
+            RevertOriginalValue(hive, RegistryConstants.SubKeys.ScreenSaver, RegistryConstants.ValueNames.ScreenSaver);
+            RevertOriginalValue(hive, RegistryConstants.SubKeys.ScreenSaverDomainPolicy, RegistryConstants.ValueNames.ScreenSaver);
+            RevertOriginalValue(hive, RegistryConstants.SubKeys.StartupProcess, RegistryConstants.ValueNames.StartupProcess);
+        }
+
+        private void RevertRegistrySettingsForDifferentUser(string securityId)
+        {
+            var hive = RegistryHive.Users;
+            RevertOriginalValue(hive, $"{securityId}\\{RegistryConstants.SubKeys.ScreenSaver}", RegistryConstants.ValueNames.ScreenSaver);
+            RevertOriginalValue(hive, $"{securityId}\\{RegistryConstants.SubKeys.ScreenSaverDomainPolicy}", RegistryConstants.ValueNames.ScreenSaver);
+            RevertOriginalValue(hive, $"{securityId}\\{RegistryConstants.SubKeys.StartupProcess}", RegistryConstants.ValueNames.StartupProcess);
+        }
+
+        private void SetValue(RegistryHive targetHive, string subKeyName, string name, string value)
+        {
+            //take backup if it exists
+            string origValue = _registryManager.GetValue(targetHive, subKeyName, name);
+            if (!string.IsNullOrEmpty(origValue))
```

Comment:
This is quite subtle. BackupAndSetValue might be a better method name. Although I still question blind backup/revert due to clobber question.